### PR TITLE
Top sites panel migration

### PR DESCRIPTION
**Pull request by iccub**
What I did
1. Copied top sites panel files until the app was compiling
2. Attached to BVC
3. Removed unused Bookmark and History panels.
4. Small code cleanup for top sites classes

I also spend some time trying to remove more code, we don't need their home panels at all, but it was tightly coupled in few places and I decided to give up and do it in another time/PR.

TODO:
- new tab doesn't have a title, this will get fixed after we move onto our `TabMO`
- DDG promo/popup
- opening private mode doesn't open new tab immediately like with v1(need to do it manually), I left it until we implement private browsing mode

zrzut ekranu 2018-06-19 o 16 30 34

### CODE REVIEW

File: Client/Frontend/Browser/HomePanel/favorites/FavoritesTileDecorator.swift [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                    setDefaultTile()
+                    return
+            }
+
+            cell.imageView.backgroundColor = website.backgroundColor
+            cell.imageView.contentMode = .scaleAspectFit
+            cell.imageView.layer.minificationFilter = kCAFilterTrilinear
+            cell.showBorder(!UIApplication.isInPrivateMode)
+
+            UIGraphicsBeginImageContextWithOptions(image.size, false, 0)
```
Comment (tmancey) [inline Client/Frontend/Browser/HomePanel/favorites/FavoritesTileDecorator.swift]: It is recommended to use UIGraphicsImageRenderer for iOS 10+

e.g.

```
extension UIImage {

    func scale(toSize size: CGSize) -> UIImage {
        if self.size == size {
            return self
        }

        let rendererFormat = UIGraphicsImageRendererFormat.default()
        rendererFormat.opaque = true
        rendererFormat.scale = 1.0

        let scaledImageRect = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        let renderer = UIGraphicsImageRenderer(size: size, format: rendererFormat)
        let scaledImage = renderer.image { context in
            self.draw(in: scaledImageRect)
        }

        return scaledImage
    }

}

```

File: Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            let m = max(maxScale / aspectWidth, maxScale / aspectHeight)
+            aspectWidth *= m
+            aspectHeight *= m
+        }
+        let aspectRatio:CGFloat = min(aspectWidth, aspectHeight)
+        scaledImageRect.size.width = image.size.width * aspectRatio
+        scaledImageRect.size.height = image.size.height * aspectRatio
+        scaledImageRect.origin.x = (size.width - scaledImageRect.size.width) / 2.0
+        scaledImageRect.origin.y = (size.height - scaledImageRect.size.height) / 2.0
+        UIGraphicsBeginImageContextWithOptions(size, false, 0)
```
Comment (tmancey) [inline Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift]: It is recommended to use UIGraphicsImageRenderer for iOS 10+

e.g.

```
extension UIImage {

    func scale(toSize size: CGSize) -> UIImage {
        if self.size == size {
            return self
        }

        let rendererFormat = UIGraphicsImageRendererFormat.default()
        rendererFormat.opaque = true
        rendererFormat.scale = 1.0

        let scaledImageRect = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        let renderer = UIGraphicsImageRenderer(size: size, format: rendererFormat)
        let scaledImage = renderer.image { context in
            self.draw(in: scaledImageRect)
        }

        return scaledImage
    }

}

```

File: Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        backgroundColor = UIColor.clear
+        textLabel.font = DynamicFontHelper.defaultHelper.DefaultSmallFont
+        textLabel.textColor = UIApplication.isInPrivateMode ? UIColor(rgb: 0xDBDBDB) : UIColor(rgb: 0x2D2D2D)
+        imageView.backgroundColor = UIColor.clear
+        imageView.image = nil
+    }
+    
+    private func updateSelectedHighlightedState() {
+        let activated = isSelected || isHighlighted
+        self.imageView.alpha = activated ? 0.7 : 1.0
```
Comment (tmancey) [inline Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift]: 0.7 and 1.0 should be constants

File: Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func prepareForReuse() {
+        super.prepareForReuse()
+        editButton.isHidden = true
+        showBorder(false)
+        backgroundColor = UIColor.clear
+        textLabel.font = DynamicFontHelper.defaultHelper.DefaultSmallFont
+        textLabel.textColor = UIApplication.isInPrivateMode ? UIColor(rgb: 0xDBDBDB) : UIColor(rgb: 0x2D2D2D)
```
Comment (tmancey) [inline Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift]: RGB values should be defined as constants in UX.swift

File: Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift (hunk 2 above)
Comment (jhreis) [reply Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift]: Filed a ticket for this. This is just directly moved code, so we can fix this up in another PR. Great find!
https://github.com/brave/brave-ios/issues/88

**Review by jhreis — APPROVED** [review]

File: Client/Frontend/Browser/HomePanel/favorites/FavoritesDataSource.swift [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+private let log = Logger.browserLogger
+
+class FavoritesDataSource: NSObject, UICollectionViewDataSource {
+    var frc: NSFetchedResultsController<NSFetchRequestResult>?
+    weak var collectionView: UICollectionView?
+
+    var isEditing: Bool = false {
+        didSet {
+            if isEditing != oldValue {
```
Comment (jhreis) [inline Client/Frontend/Browser/HomePanel/favorites/FavoritesDataSource.swift:17]: Replace with early return:

```
if isEditing == oldValue { return }
let name = ...
```

**Comment by jhreis** [comment]
Just clean up the conflicts and the one comment I left, and should be good to go. Needs to be merged with #79 though.

File: Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift (hunk 2 above)
Comment (iccub) [reply Client/Frontend/Browser/HomePanel/favorites/FavoriteCell.swift]: I added the new resize method already but forgot to remove the old one, sorry.

**Review by tmancey — APPROVED** [review]

**Pull request merged**